[docs] Update the documentation to include a description of the framework's window management capabilities (#5363)
DevExpress/testcafe: docs/articles/documentation/guides/advanced-guides/multiple-browser-windows.md

---
layout: docs
title: Multiple Browser Windows
permalink: /documentation/guides/advanced-guides/multiple-browser-windows.html
---
# Multiple Browser Windows

The TestCafe API includes methods that open, close, and switch between browser windows. You can test websites with pop-up windows and OAuth login forms, debug complex multi-window applications, or run multiple instances of the same web app side by side.

> >This is a beta feature. Browser support is limited to local instances of Chrome and Firefox. Videos and screenshots of child windows cannot be captured. The available functionality is subject to further revisions. Use at your own risk.

Reviewer: > Use at your own risk.

We were just going to say that we don't recommend production usage. There's no reason to talk about risks.

## Handle client-side window events

When your page launches a new window, the test automatically continues in the newly open window. When that window is closed, the test switches back to its parent.

```js
import { Selector } from 'testcafe';

fixture `Login page`
    .page('https://login.wrike.com/login/');

const googleButton = Selector('div.login-panel-footer__login-with > button');

test('Login via Google', async t => {
    await t
        .click(googleButton)
        .typeText('input[type=email]', 'This text will be entered inside the login dialog');
});
```

## Open a new window

Use the [t.openWindow](../../reference/test-api/testcontroller/openwindow.md) method to open a new window:

```js
import { Selector, ClientFunction } from 'testcafe';

fixture `Test page`
    .page('https://devexpress.github.io/testcafe/example/');

test('Open a new window', async t => {
    await t.openWindow('http://example.com');

    const url = await t.eval(() => document.documentURI);
    await t.expect(url).eql('http://example.com/');
});
```

> >All open browser windows share the same cookies and user storage. Support for per-window user sessions is not yet available.

Reviewer: First, 'not yet' looks like a promise. Second (which I didn't say the last time), the documentation usually doesn't name features that aren't yet available. This limitation could be rephrased without mentioning that something specific is not yet here. For example:

*All open browser windows share the same cookies, user storage, and, therefore, the same user session.*

---
layout: docs
title: Multiple Browser Windows
permalink: /documentation/guides/advanced-guides/multiple-browser-windows.html
---
# Multiple Browser Windows

> The TestCafe API includes methods that open, close, and switch between browser windows. You can test websites with pop-up windows and OAuth login forms, debug complex multi-window applications, or run multiple instances of the same web app side by side.

Reviewer 2: ```suggestion
The TestCafe API includes methods that open, close, and switch between browser windows. You can test websites with pop-up windows and OAuth login forms, debug complex multi-window applications, or run multiple instances of the same web app side-by-side.
```

> ⚠ This is a **beta** feature. Not recommended for use in production environments. Browser support is limited to local instances of Chrome and Firefox. Videos and screenshots of child windows cannot be captured. The available functionality is subject to further revisions.

Reviewer: > Not recommended for use in production environments.

Please use full sentences.

> ️🛠️ Use the `--disable-multiple-windows` CLI flag to disable this feature if you encounter compatibility issues with legacy tests.

Reviewer: Please see this comment: https://github.com/DevExpress/testcafe/pull/5363#discussion_r463606796

When we say 'legacy tests' between ourselves, we mean specifically tests written for TestCafe v2015.1, and that's how they should be called in the docs.

fixture `Login page`
    .page('https://login.wrike.com/login/');

const googleButton = Selector('div.login-panel-footer__login-with > button');

test('Login via Google', async t => {
    await t
        .click(googleButton)
        .typeText('input[type=email]', 'This text will be entered inside the login dialog');
});
```

## Open a new window

Use the [t.openWindow](../../reference/test-api/testcontroller/openwindow.md) method to open a new window:

```js
import { Selector, ClientFunction } from 'testcafe';

fixture `Test page`
    .page('https://devexpress.github.io/testcafe/example/');

test('Open a new window', async t => {
    await t.openWindow('http://example.com');

    const url = await t.eval(() => document.documentURI);
    await t.expect(url).eql('http://example.com/');
});
```

>All open browser windows share a single pool of client-side storage. Only one [user role](https://devexpress.github.io/testcafe/documentation/guides/advanced-guides/authentication.html#user-roles) can be active at a time.

## Switch between windows

The [t.switchToWindow](../../reference/test-api/testcontroller/switchtowindow.md) method lets you switch between browser windows.

> It takes an instance of the Window object as an argument:

Reviewer: So what was our final decision regarding "window object"? Was it "window descriptor"?

I'm okay with that, although "window identifier" seems slightly better to me.

Reviewer: There can also be no "instances of objects" BTW. You can create a class instance in some languages, but here we should just say "object" (identifier/descriptor).

> It takes a window descriptor as an argument:

Reviewer 3: It what? specify

Reviewer 2: Do you mean 
```suggestion
This method takes a window descriptor as an argument:
```

```js
fixture `Example page`
    .page('http://example.com');

test('Switch to a specific window', async t => {
    const initialWindow = await t.getCurrentWindow();
    const window2 = await t.openWindow('http://example1.com');

>     const window3 = await t.openWindow('http://example2.com');

Reviewer: `http://example1.com` no longer opens for me, and this test does not work. Maybe we change them to `devexpress.com` or `github.com`?

    await t.switchToWindow(initialWindow);

    const url = await t.eval(() => document.documentURI);
    await t.expect(url).eql('http://example.com/');
});
```

> You can also pass [a predicate](../../reference/test-api/testcontroller/switchtowindow.md#tswitchtowindowpredicate) with the description of the window you need:

Reviewer 2: ```suggestion
You can also pass [a predicate](../../reference/test-api/testcontroller/switchtowindow.md#tswitchtowindowpredicate) with the description of the window:
```

```js
await t.switchToWindow(w => w.title === 'Example Domain' && w.url.host === 'example.com');
```

Use the [t.switchToParentWindow](../../reference/test-api/testcontroller/switchtoparentwindow.md) method to access the window's parent:

```js
fixture `Example page`
    .page('http://www.example.com/');

test('Switch to a parent window', async t => {
    await t
        .openWindow('https://devexpress.com')
        .switchToParentWindow();

    const url = await t.eval(() => document.documentURI);
    await t.expect(url).eql('http://www.example.com/');
});
```

Use the [t.switchToPreviousWindow](../../reference/test-api/testcontroller/switchtopreviouswindow.md) method to access the second to last open window:

```js
test('Switch back', async t => {
    await t
        .openWindow('http://www.example.com')
        .openWindow('https://devexpress.com');
        .switchToPreviousWindow();

    const url = await t.eval(() => document.documentURI);
    await t.expect(url).eql('http://www.example.com/');
});
```

## Close an existing window

>You cannot close windows with open children.

> Call the [t.closeWindow](../../reference/test-api/testcontroller/closewindow.md) method without parameters to close the currently active window:

Reviewer 2: ```suggestion
Call the [t.closeWindow](../../reference/test-api/testcontroller/closewindow.md) method without parameters to close the active window:
```

```js
fixture `Example page`
    .page('http://www.example.com');

test('Close the current window', async t => {
    const window1 = await t.openWindow('http://devexpress.com');
    await t.closeWindow();

    const url = await t.eval(() => document.documentURI);
    await t.expect(url).eql('http://www.example.com/');
    });

```

> Pass a window descriptor when you call the [t.closeWindow](../../reference/test-api/testcontroller/closewindow.md) method to close a particular window:

Reviewer 3: ```suggestion
Pass a window descriptor when you call the [t.closeWindow](../../reference/test-api/testcontroller/closewindow.md) method to close a window:
```

Author: I was trying to highlight the difference between what the t.closeWindow method does when called without arguments (it closes the currently active window) and what it does with a window descriptor present. Would `Pass a window descriptor when you call the [t.closeWindow](../../reference/test-api/testcontroller/closewindow.md) to close the corresponding window` work?

---
layout: docs
title: Multiple Browser Windows
permalink: /documentation/guides/advanced-guides/multiple-browser-windows.html
---
# Multiple Browser Windows

The TestCafe API includes methods that open, close, and switch between browser windows. You can test websites with pop-up windows and OAuth login forms, debug complex multi-window applications, or run multiple instances of the same web app side-by-side.

> ⚠ This is a **beta** feature. Browser support is limited to local instances of Chrome and Firefox. Videos and screenshots of child windows cannot be captured. The available functionality is subject to further revisions. Please refrain from using this feature in production environments.

Reviewer 2: ```suggestion
⚠ This is a **beta** feature. Browser support is limited to local instances of Chrome and Firefox. You cannot capture videos or screenshots of child windows. The available functionality is subject to further revisions. Please do not use this feature in production environments.
```

️🛠️ Use the `--disable-multiple-windows` CLI flag to disable support for multiple browser windows if you encounter compatibility issues with [TestCafe v2015.1](http://testcafe.devexpress.com/documentation) tests.

## Handle client-side window events

> When your page launches a new window, the test automatically continues in the newly open window. When that window is closed, the test switches back to its parent.

Reviewer 2: ```suggestion
When your page launches a new window, the test automatically continues in the newly opened window. When that window is closed, the test switches back to its parent.
```

```js
import { Selector } from 'testcafe';

fixture `Login page`
    .page('https://login.wrike.com/login/');

const googleButton = Selector('div.login-panel-footer__login-with > button');

test('Login via Google', async t => {
    await t
        .click(googleButton)
        .typeText('input[type=email]', 'This text will be entered inside the login dialog');
});
```

## Open a new window

Use the [t.openWindow](../../reference/test-api/testcontroller/openwindow.md) method to open a new window:

```js
import { Selector, ClientFunction } from 'testcafe';

fixture `Test page`
    .page('https://devexpress.github.io/testcafe/example/');

test('Open a new window', async t => {
    await t.openWindow('http://example.com');

    const url = await t.eval(() => document.documentURI);
    await t.expect(url).eql('http://example.com/');
});
```

> >All open browser windows share a single pool of client-side storage. Only one [user role](https://devexpress.github.io/testcafe/documentation/guides/advanced-guides/authentication.html#user-roles) can be active at a time.

Reviewer 2: ```suggestion
>All browser windows share the client-side storage. Only one [user role](https://devexpress.github.io/testcafe/documentation/guides/advanced-guides/authentication.html#user-roles) can be active at a time.
```

## Obtain window descriptors

> Window descriptors are objects that reference individual browser windows. Window descriptors are useful if you have more than two windows open at once, and want to easily distinguish between them.

Reviewer 2: ```suggestion
Window descriptors are objects that reference individual browser windows. Window descriptors can be used to distinguish between different windows.
```

The [t.openWindow](../../reference/test-api/testcontroller/openwindow.md) method returns a window descriptor for the newly open window.

> The [t.getCurrentWindow](../../reference/test-api/testcontroller/getcurrentwindow.md) method returns a window descriptor for the active window. Use it to obtain descriptors for windows that open without a [t.openWindow](../../reference/test-api/testcontroller/openwindow.md) declaration, such as the initial browser window.

Reviewer 2: ```suggestion
The [t.getCurrentWindow](../../reference/test-api/testcontroller/getcurrentwindow.md) method returns a window descriptor for the active window. Use it to obtain descriptors for windows that open without a [t.openWindow](../../reference/test-api/testcontroller/openwindow.md) declaration, for example, the initial browser window.
```

```js
fixture `Example page`
    .page('http://example.com');

test('Generate window descriptors', async t => {
    const initialWindow = await t.getCurrentWindow();
    const window2 = await t.openWindow('http://devexpress.com');
    });
```

## Switch between windows

> The [t.switchToWindow](../../reference/test-api/testcontroller/switchtowindow.md) method lets you switch between browser windows.

Reviewer 2: ```suggestion
The [t.switchToWindow](../../reference/test-api/testcontroller/switchtowindow.md) method allows you to switch between browser windows.
```

```js
fixture `Example page`
    .page('http://example.com');

test('Switch to a specific window', async t => {
    const initialWindow = await t.getCurrentWindow();
    const window2 = await t.openWindow('http://devexpress.com');
    const window3 = await t.openWindow('http://github.com');

    await t.switchToWindow(initialWindow);

    const url = await t.eval(() => document.documentURI);
    await t.expect(url).eql('http://example.com/');
});
```

You can also pass [a predicate](../../reference/test-api/testcontroller/switchtowindow.md#tswitchtowindowpredicate) with a description of the window:

```js
await t.switchToWindow(w => w.title === 'Example Domain' && w.url.host === 'example.com');
```

Use the [t.switchToParentWindow](../../reference/test-api/testcontroller/switchtoparentwindow.md) method to access the window's parent:

```js
fixture `Example page`
    .page('http://www.example.com/');

test('Switch to a parent window', async t => {
    await t
        .openWindow('https://devexpress.com')
        .switchToParentWindow();

    const url = await t.eval(() => document.documentURI);
    await t.expect(url).eql('http://www.example.com/');
});
```

> Use the [t.switchToPreviousWindow](../../reference/test-api/testcontroller/switchtopreviouswindow.md) method to access the second to last open window:

Reviewer 2: How does "switchToPreviousWindow" relate to "the second to last"?
